### PR TITLE
Add proper path fallback to cache entries

### PR DESCRIPTION
**Pull request by juliusknorr**
This fixes two issues when sharing a complete group folder:

Search:
- Create a group folder and add some files to it
- Share the whole group folder with another user or guest account
- Search in the home folder for a file that exists in the groupfolder as user or guest
:boom: Filename not displayed properly and path is set wrong
:heavy_check_mark: File is displayed and linked properly

Activity:
- Create a group folder and add some files to it
- As admin share the whole group folder with a guest account
- As a guest user upload a file to the group folder
:boom: The activity app for the admin shows a broken entry
:heavy_check_mark: The activity app for the admin shows the correct file entry

### CODE REVIEW

File: apps/files_sharing/lib/Cache.php [hunk 1]
```diff
@@ -75,6 +75,10 @@ protected function getRoot() {
 		if (is_null($this->root)) {
 			$absoluteRoot = $this->sourceRootInfo->getPath();
 
+			if ($this->sourceRootInfo->getMimeType() === 'httpd/unix-directory') {
+				$absoluteRoot = rtrim($absoluteRoot, '/') . '/';
```
Comment (juliushaertl) [inline apps/files_sharing/lib/Cache.php]: Hm, the tests don't seem to be too happy. When just applying the the normalization if we are inside a jail storage the tests run fine. But looking the failing tests it also feels like this is fixing a symptom as well :wink:

Comment (juliushaertl) [reply apps/files_sharing/lib/Cache.php]: Shouldn't we just properly normalize paths before checking that path is in root https://github.com/nextcloud/server/blob/9b3cc72f7c033c4dba9a2b1e21e0e38f488318b5/lib/private/Files/Storage/Wrapper/Jail.php#L68

Comment (juliushaertl) [reply apps/files_sharing/lib/Cache.php]: @rullzer It seems that the existing Jail/CacheJail logic generally assumes that the root is a string, while the FilesSharing\Cache initializes it with null. That fix would also solve the issue:

```diff
diff --git a/apps/files_sharing/lib/Cache.php b/apps/files_sharing/lib/Cache.php
index 4dbe4b75b2..b0575a4a6b 100644
--- a/apps/files_sharing/lib/Cache.php
+++ b/apps/files_sharing/lib/Cache.php
@@ -67,7 +67,7 @@ class Cache extends CacheJail {
 
                parent::__construct(
                        null,
-                       null
+                       ''
                );
        }
```

Comment (rullzer) [reply apps/files_sharing/lib/Cache.php]: ooo then lets do that!

Comment (juliushaertl) [reply apps/files_sharing/lib/Cache.php]: https://github.com/nextcloud/server/pull/16985

**Pull request closed without merging**